File: src/rooftop-cat/weather.js
/* weather.js — ambience & weather effects
   Modes: none | rain | snow | fog | storm
   Exports:
     - initWeather(mode, state, canvas, reduceMotion)
     - advanceLightning(state, dt, canvas)
     - renderLightning(ctx, state, canvas)
*/

function viewSize(canvas){
  // Map backing store size → CSS pixels using the effective DPR
  const cssW = canvas.clientWidth || Math.max(1, Math.round(canvas.width / (window.devicePixelRatio || 1)));
  const effDpr = Math.max(1, canvas.width / cssW);
  const w = Math.floor(canvas.width / effDpr);
  const h = Math.floor(canvas.height / effDpr);
  return { w, h, dpr: effDpr };
}

export function initWeather(mode, state, canvas, reduceMotion){
  state.rain = [];
  state.snow = [];
  state.fog  = [];
  state.fogTex = null;

  // lightning / storm state
  state.lightning = [];
  state.storm = null;

  if (mode === "rain")       initRain(state, canvas, reduceMotion);
  else if (mode === "snow")  initSnow(state, canvas, reduceMotion);
  else if (mode === "fog")   initFog(state, canvas, reduceMotion);
  else if (mode === "storm") initStorm(state, canvas, reduceMotion);
}

/* ---------------- Rain / Snow / Fog ---------------- */

function initRain(state, canvas, reduceMotion){
  const { w, h } = viewSize(canvas);
  const overscan = Math.floor(Math.max(80, w * 0.08));

  // Keep original feel: lighter rain with depth-scaled slant/speed
  // Count scales with area but tracks the original 80/160 around 1280x720.
  const baselineArea = 1280 * 720;
  const k = (w * h) / baselineArea;
  const baseCount = reduceMotion ? 80 : 160;
  const count = Math.max(30, Math.floor(baseCount * k));

  state.rain.length = 0;
  for (let i = 0; i < count; i++){
    const z = Math.random();
    state.rain.push({
      x: Math.random() * (w + overscan) - overscan,
      y: Math.random() * (h + 120) - 120,
      vx: -40 - 40 * z,
      vy: 320 + 380 * z,
      z
    });
  }
}

function initSnow(state, canvas, reduceMotion){
  const { w, h } = viewSize(canvas);
  const overscan = Math.floor(Math.max(80, w * 0.08));

  // Original drift/sway, count scales with area around 80/140 baseline.
  const baselineArea = 1280 * 720;
  const k = (w * h) / baselineArea;
  const baseCount = reduceMotion ? 80 : 140;
  const count = Math.max(20, Math.floor(baseCount * k));

  state.snow.length = 0;
  for (let i = 0; i < count; i++){
    const size = 1 + Math.random() * 2.2;
    state.snow.push({
      x: Math.random() * (w + overscan) - overscan,
      y: Math.random() * (h + 80) - 80,
      vx: -14 - Math.random() * 10,
      vy: 18 + Math.random() * 28,
      sway: Math.random() * Math.PI * 2,
      swaySpeed: 0.6 + Math.random() * 1.2,
      r: size
    });
  }
}

function makeFogTexture(){
  // Bigger, softer fog (your working version)
  const c = document.createElement("canvas");
  c.width = c.height = 256;
  const g = c.getContext("2d");
  const grad = g.createRadialGradient(128,128,20, 128,128,128);
  grad.addColorStop(0.00, "rgba(220,235,255,0.30)");
  grad.addColorStop(0.50, "rgba(220,235,255,0.20)");
  grad.addColorStop(1.00, "rgba(220,235,255,0.00)");
  g.fillStyle = grad;
  g.fillRect(0,0,256,256);
  return c;
}

function initFog(state, canvas, reduceMotion){
  const { w } = viewSize(canvas);
  const overscan = Math.floor(Math.max(120, w * 0.12));
  const yTop = Math.max(0, state.groundY - 170);
  const yBottom = state.groundY + 36;
  const fogBandH = Math.max(100, yBottom - yTop);

  const count = Math.max(6, Math.floor(w / 140)); // big, soft blobs
  state.fogTex = makeFogTexture();
  state.fog.length = 0;

  for (let i = 0; i < count; i++){
    const r = 80 + Math.random() * 160;
    state.fog.push({
      x: Math.random() * (w + overscan) - overscan * 0.5,
      y: yTop + Math.random() * fogBandH,
      r,
      vx: -20 - Math.random() * 25,
      a: 0.35 + Math.random() * 0.25,
      phi: Math.random() * Math.PI * 2,
      swaySpeed: 0.8 + Math.random() * 1.2
    });
  }
}

/* ---------------- Storm (heavier rain + original lightning) ---------------- */

function initStorm(state, canvas, reduceMotion){
  const { w, h } = viewSize(canvas);
  const overscan = Math.floor(Math.max(80, w * 0.08));

  // Heavier rain than initRain, with original storm speeds.
  const baselineArea = 1280 * 720;
  const k = (w * h) / baselineArea;
  const baseCount = reduceMotion ? 180 : 320; // > rain
  const count = Math.max(60, Math.floor(baseCount * k));

  state.rain.length = 0;
  for (let i = 0; i < count; i++){
    const z = Math.random();
    state.rain.push({
      x: Math.random() * (w + overscan) - overscan,
      y: Math.random() * (h + 120) - 120,
      vx: -80 - 70 * z,      // stronger wind
      vy: 420 + 520 * z,     // faster drops
      z
    });
  }

  // Lightning state (original schedule/flash)
  state.lightning = [];
  state.storm = {
    rm: !!reduceMotion,
    flash: 0,
    nextBolt: (reduceMotion ? 3.5 : 3.0) + Math.random() * (reduceMotion ? 4.5 : 3.5),
  };
}

/* ---------------- Lightning (original look & timing) ---------------- */

export function advanceLightning(state, dt, canvas){
  if (!state.storm) return;

  // spawn schedule
  state.storm.nextBolt -= dt;
  if (state.storm.nextBolt <= 0){
    spawnLightning(state, canvas);
    const minGap = state.storm.rm ? 3.0 : 2.5;
    const maxGap = state.storm.rm ? 6.0 : 5.0;
    state.storm.nextBolt = minGap + Math.random() * (maxGap - minGap);
  }

  // fade flash
  if (state.storm.flash > 0){
    state.storm.flash = Math.max(0, state.storm.flash - dt * 0.6);
  }

  // age + cull
  for (const b of state.lightning) b.age += dt;
  while (state.lightning.length && state.lightning[0].age > state.lightning[0].life){
    state.lightning.shift();
  }
}

export function renderLightning(ctx, state, canvas){
  if (!state.storm) return;

  const { w, h } = viewSize(canvas);

  // bolts
  if (state.lightning.length){
    ctx.save();
    ctx.globalCompositeOperation = "lighter";

    for (const bolt of state.lightning){
      const k = 1 - (bolt.age / bolt.life);
      const jitter = 0.75 + Math.random() * 0.25;

      // outer glow
      ctx.globalAlpha = 0.55 * k * jitter;
      ctx.strokeStyle = "#8bd3ff";
      ctx.lineWidth = 6;
      strokePath(ctx, bolt.pts);

      // core
      ctx.globalAlpha = 0.9 * k;
      ctx.strokeStyle = "#eef7ff";
      ctx.lineWidth = 2.4;
      strokePath(ctx, bolt.pts);

      // branches
      ctx.globalAlpha = 0.85 * k;
      ctx.lineWidth = 1.8;
      for (const br of bolt.branches) strokePath(ctx, br);
    }

    ctx.restore();
  }

  // screen flash
  if (state.storm.flash > 0){
    ctx.save();
    ctx.globalAlpha = Math.min(0.45, state.storm.flash);
    ctx.fillStyle = "rgba(180,205,245,0.90)";
    ctx.fillRect(0, 0, w, h);
    ctx.restore();
  }
}

function strokePath(ctx, pts){
  if (!pts || pts.length < 2) return;
  ctx.beginPath();
  ctx.moveTo(pts[0].x, pts[0].y);
  for (let i = 1; i < pts.length; i++) ctx.lineTo(pts[i].x, pts[i].y);
  ctx.stroke();
}

function spawnLightning(state, canvas){
  const { w } = viewSize(canvas);
  const gy = state.groundY;

  // Original path: from high sky down toward mid-sky above buildings
  const x0 = 40 + Math.random() * (w - 80);
  const y0 = 30 + Math.random() * 120;
  const y1 = gy - (140 + Math.random() * 100);
  const steps = 10 + Math.floor(Math.random() * 5);

  const pts = [];
  let x = x0;
  for (let i = 0; i <= steps; i++){
    const t = i / steps;
    const ny = y0 + (y1 - y0) * t + (Math.random() * 8 - 4);
    const jitter = (1 - t) * 28; // bigger sideways near top
    x += (Math.random() * 2 - 1) * jitter * 0.6;
    pts.push({ x, y: ny });
  }

  // small downward branches
  const branches = [];
  for (let i = 2; i < pts.length - 2; i++){
    if (Math.random() < 0.25){
      const b = [pts[i]];
      let bx = pts[i].x, by = pts[i].y;
      const len = 2 + Math.floor(Math.random() * 3);
      for (let k = 0; k < len; k++){
        bx += (Math.random() * 14 - 7);
        by += (Math.random() * 20 + 10);
        b.push({ x: bx, y: by });
      }
      branches.push(b);
    }
  }

  state.lightning.push({ pts, branches, life: 0.22, age: 0 });
  state.storm.flash = Math.max(state.storm.flash, 0.35);
}

File: python setup/weather.py

# weather.py — ambience & weather effects
# Modes: none | rain | snow | fog | storm
# Exports:
#   initWeather(mode, state, canvas, reduceMotion)
#   advanceLightning(state, dt, canvas)
#   renderLightning(ctx, state, canvas)
#
# Notes:
# - 'canvas' may be a pygame.Surface, a dict with width/height/dpr, an (w,h) tuple,
#   or just a width (int). We map to CSS-like pixels similarly to the JS viewSize().

import math
import random
import pygame

__all__ = ["initWeather", "advanceLightning", "renderLightning"]

_FOG_TEX = None

def _get_fog_texture():
    global _FOG_TEX
    if _FOG_TEX is None:
        _FOG_TEX = _make_fog_texture()
    return _FOG_TEX

# ---------------- view size helper ----------------

def _view_size(canvas):
    # Returns (w, h, dpr)
    # pygame.Surface
    try:
        w = int(canvas.get_width())
        h = int(canvas.get_height())
        return w, h, 1.0
    except AttributeError:
        pass

    # dict with width/height/dpr
    if isinstance(canvas, dict):
        cssW = int(canvas.get("clientWidth") or max(1, round(canvas["width"] / max(1.0, float(canvas.get("dpr", 1.0))))))
        effDpr = max(1.0, float(canvas["width"]) / cssW) if cssW else 1.0
        w = int(canvas["width"] / effDpr)
        h = int(canvas.get("height", 0) / effDpr)
        return w, h, effDpr

    # (w,h) tuple
    if isinstance(canvas, (tuple, list)) and len(canvas) >= 2:
        return int(canvas[0]), int(canvas[1]), 1.0

    # single width
    if isinstance(canvas, (int, float)):
        w = int(canvas)
        h = 720
        return w, h, 1.0

    # fallback
    return 1280, 720, 1.0

# ---------------- API ----------------

def initWeather(mode, state, canvas, reduceMotion=False):
    state["rain"] = []
    state["snow"] = []
    state["fog"]  = []
    state["fogTex"] = None

    # lightning / storm state
    state["lightning"] = []
    state["storm"] = None

    if mode == "rain":
        _init_rain(state, canvas, reduceMotion)
    elif mode == "snow":
        _init_snow(state, canvas, reduceMotion)
    elif mode == "fog":
        _init_fog(state, canvas, reduceMotion)
    elif mode == "storm":
        _init_storm(state, canvas, reduceMotion)
    # else: none

# ---------------- Rain / Snow / Fog ----------------

def _init_rain(state, canvas, reduceMotion):
    w, h, _ = _view_size(canvas)
    overscan = int(max(80, w * 0.08))

    baselineArea = 1280 * 720
    k = (w * h) / baselineArea
    baseCount = 80 if reduceMotion else 160
    count = max(30, int(baseCount * k))

    state["rain"].clear()
    for _ in range(count):
        z = random.random()
        state["rain"].append({
            "x": random.random() * (w + overscan) - overscan,
            "y": random.random() * (h + 120) - 120,
            "vx": -40 - 40 * z,
            "vy": 320 + 380 * z,
            "z": z
        })

def _init_snow(state, canvas, reduceMotion):
    w, h, _ = _view_size(canvas)
    overscan = int(max(80, w * 0.08))

    baselineArea = 1280 * 720
    k = (w * h) / baselineArea
    baseCount = 80 if reduceMotion else 140
    count = max(20, int(baseCount * k))

    state["snow"].clear()
    for _ in range(count):
        size = 1 + random.random() * 2.2
        state["snow"].append({
            "x": random.random() * (w + overscan) - overscan,
            "y": random.random() * (h + 80) - 80,
            "vx": -14 - random.random() * 10,
            "vy": 18 + random.random() * 28,
            "sway": random.random() * math.pi * 2,
            "swaySpeed": 0.6 + random.random() * 1.2,
            "r": size
        })

def _make_fog_texture():
    # Build a 256x256 radial gradient surface like the Canvas version
    c = pygame.Surface((256, 256), pygame.SRCALPHA)
    cx = cy = 128
    max_r = 128
    # Approx radial gradient by drawing concentric circles from center outward
    # Stops: 0.0: 0.30, 0.50: 0.20, 1.0: 0.00
    for i in range(max_r, 0, -1):
        t = i / max_r
        # piecewise lerp of alpha across stops
        if t >= 0.5:
            # between 0.5 .. 1.0 -> alpha 0.20 .. 0.00
            a = (t - 0.5) / 0.5
            alpha = (0.20 * (1 - a) + 0.00 * a)
        else:
            # between 0.0 .. 0.5 -> alpha 0.30 .. 0.20
            a = t / 0.5
            alpha = (0.30 * a + 0.20 * (1 - a))
        col = (220,235,255, int(alpha * 255))
        pygame.draw.circle(c, col, (cx, cy), i)
    return c

def _init_fog(state, canvas, reduceMotion):
    w, h, _ = _view_size(canvas)
    overscan = int(max(120, w * 0.12))
    yTop = max(0, int(state.get("groundY", h - 220) - 170))
    yBottom = int(state.get("groundY", h - 220) + 36)
    fogBandH = max(100, yBottom - yTop)

    count = max(6, int(w / 140))
    state["fogTex"] = _get_fog_texture()
    state["fog"].clear()

    for _ in range(count):
        r = 80 + random.random() * 160
        state["fog"].append({
            "x": random.random() * (w + overscan) - overscan * 0.5,
            "y": yTop + random.random() * fogBandH,
            "r": r,
            "vx": -20 - random.random() * 25,
            "a": 0.35 + random.random() * 0.25,
            "phi": random.random() * math.pi * 2,
            "swaySpeed": 0.8 + random.random() * 1.2
        })

# ---------------- Storm (heavier rain + lightning) ----------------

def _init_storm(state, canvas, reduceMotion):
    w, h, _ = _view_size(canvas)
    overscan = int(max(80, w * 0.08))

    baselineArea = 1280 * 720
    k = (w * h) / baselineArea
    baseCount = 180 if reduceMotion else 320
    count = max(60, int(baseCount * k))

    state["rain"].clear()
    for _ in range(count):
        z = random.random()
        state["rain"].append({
            "x": random.random() * (w + overscan) - overscan,
            "y": random.random() * (h + 120) - 120,
            "vx": -80 - 70 * z,
            "vy": 420 + 520 * z,
            "z": z
        })

    state["lightning"] = []
    state["storm"] = {
        "rm": bool(reduceMotion),
        "flash": 0.0,
        "nextBolt": (3.5 if reduceMotion else 3.0) + random.random() * (4.5 if reduceMotion else 3.5),
    }

# ---------------- Lightning (schedule + drawing) ----------------

def advanceLightning(state, dt, canvas):
    if not state.get("storm"):
        return

    # spawn schedule
    state["storm"]["nextBolt"] -= dt
    if state["storm"]["nextBolt"] <= 0:
        _spawn_lightning(state, canvas)
        minGap = 3.0 if state["storm"]["rm"] else 2.5
        maxGap = 6.0 if state["storm"]["rm"] else 5.0
        state["storm"]["nextBolt"] = minGap + random.random() * (maxGap - minGap)

    # fade flash
    if state["storm"]["flash"] > 0:
        state["storm"]["flash"] = max(0.0, state["storm"]["flash"] - dt * 0.6)

    # age + cull
    for b in state["lightning"]:
        b["age"] += dt
    while state["lightning"] and state["lightning"][0]["age"] > state["lightning"][0]["life"]:
        state["lightning"].pop(0)

def renderLightning(ctx, state, canvas):
    if not state.get("storm"):
        return

    w, h, _ = _view_size(canvas)
    W, H = int(w), int(h)

    # --- reuse bolt layer
    layer = state.get("_stormLayer")
    if layer is None or layer.get_width() != W or layer.get_height() != H:
        layer = pygame.Surface((W, H), pygame.SRCALPHA)
        state["_stormLayer"] = layer
    else:
        layer.fill((0, 0, 0, 0))  # clear

    # bolts
    if state["lightning"]:
        for bolt in state["lightning"]:
            k = 1.0 - (bolt["age"] / bolt["life"])
            jitter = 0.75 + random.random() * 0.25
            _stroke_path(layer, bolt["pts"], color=(0x8b,0xd3,0xff, int(0.55 * k * jitter * 255)), width=6)
            _stroke_path(layer, bolt["pts"], color=(0xee,0xf7,0xff, int(0.90 * k * 255)), width=3)
            for br in bolt["branches"]:
                _stroke_path(layer, br, color=(0xee,0xf7,0xff, int(0.85 * k * 255)), width=2)

        ctx.blit(layer, (0, 0))

    # --- reuse flash layer
    if state["storm"]["flash"] > 0:
        flash = state.get("_stormFlash")
        if flash is None or flash.get_width() != W or flash.get_height() != H:
            flash = pygame.Surface((W, H), pygame.SRCALPHA)
            state["_stormFlash"] = flash
        else:
            flash.fill((0, 0, 0, 0))
        flash.fill((180, 205, 245, int(min(0.45, state["storm"]["flash"]) * 255)))
        ctx.blit(flash, (0, 0))

# ---------------- internals ----------------

def _stroke_path(surface, pts, color=(255,255,255,255), width=2):
    if not pts or len(pts) < 2:
        return
    # draw to a temp small surface for per-primitive alpha
    col = color
    # Convert to integer points
    ipts = [(int(p[0]), int(p[1])) for p in pts]
    # pygame can draw lines with alpha directly on SRCALPHA surface
    pygame.draw.lines(surface, col, False, ipts, int(max(1, width)))

def _spawn_lightning(state, canvas):
    w, h, _ = _view_size(canvas)
    gy = state.get("groundY", int(h * 0.66))

    x0 = 40 + random.random() * (w - 80)
    y0 = 30 + random.random() * 120
    y1 = gy - (140 + random.random() * 100)
    steps = 10 + int(random.random() * 5)

    pts = []
    x = x0
    for i in range(steps + 1):
        t = i / steps
        ny = y0 + (y1 - y0) * t + (random.random() * 8 - 4)
        jitter = (1 - t) * 28
        x += (random.random() * 2 - 1) * jitter * 0.6
        pts.append((x, ny))

    # small downward branches
    branches = []
    for i in range(2, len(pts) - 2):
        if random.random() < 0.25:
            b = [pts[i]]
            bx, by = pts[i]
            length = 2 + int(random.random() * 3)
            for _ in range(length):
                bx += (random.random() * 14 - 7)
                by += (random.random() * 20 + 10)
                b.append((bx, by))
            branches.append(b)

    state["lightning"].append({"pts": pts, "branches": branches, "life": 0.22, "age": 0.0})
    state["storm"]["flash"] = max(state["storm"]["flash"], 0.35)
